src/mozilla_sec_eia/utils.py

```diff
@@ -78,12 +97,126 @@ def getconn() -> pg8000.dbapi.Connection:
+            composite.paste(image, (0, y))
+            y += image.height
+        return composite
+
+    def get_filing(
+        self,
+        filing: pd.Series,
+        cache_directory: Path = Path("./sec10k_filings"),
+    ) -> Path:
+        """Cache a single filing and return path."""
+        # Create cache directory
+        cache_directory.mkdir(parents=True, exist_ok=True)
+
+        blob = self.get_blob(filing["year_quarter"], filing["Filename"])
+        local_path = self._get_local_path(cache_directory, filing)
+
+        if exists := local_path.exists():
+            blob.update()
+            local_hash = _compute_md5(local_path)
+            remote_hash = blob.md5_hash
+            refresh = remote_hash != local_hash
+
+        if (not exists) or refresh:
+            blob.download_to_filename(local_path)
+
+        return local_path
+
+    def get_filings(
+        self,
+        filing_selection: pd.DataFrame,
+        cache_directory: Path = Path("./sec10k_filings"),
+    ) -> typing.Generator[Path, None, None]:
+        """Caches filings locally for quick access, and returns generator producing paths.
+
+        Args:
+            filing_selection: Pandas dataframe with same schema as metadata df where each row
+                is a filing to return.
+            cache_directory: Path to directory where filings should be cached.
+        """
+        for _, filing in filing_selection.iterrows():
+            yield self.get_filing(filing, cache_directory)
+
+    def get_ex_21_from_filing(
+        self,
+        filing: Path,
+    ) -> Image:
+        """Extract exhibit 21 from raw filing and return PIL Image of table."""
+        ex_21_pat = re.compile(
+            r"<DOCUMENT>\s?<TYPE>EX-21(\.\d)?([\S\s])*?(</DOCUMENT>)"
+        )
+        with filing.open() as f:
+            ex_21_text = ex_21_pat.search(f.read()).group(0)
+
+        pdf_bytes = io.BytesIO()
+        res = pisa.CreatePDF(ex_21_text, pdf_bytes)
+        if res.err:
+            raise RuntimeError(
+                f"Failed to create PDF from filing {str(filing)} exhibit 21."
+            )
+
+        doc = fitz.open(stream=pdf_bytes, filetype="pdf")
+        pages = [
+            Image.open(
+                io.BytesIO(doc.load_page(i).get_pixmap().pil_tobytes(format="PNG"))
+            )
+            for i in range(doc.page_count)
+        ]
+        ex_21_img = self._v_stack_images(*pages)
+
+        # Crop white border
+        nonwhite_positions = [
+            (x, y)
+            for x in range(ex_21_img.size[0])
+            for y in range(ex_21_img.size[1])
+            if ex_21_img.getdata()[x + y * ex_21_img.size[0]] != (255, 255, 255)
+        ]
+        rect = (
+            min([x for x, y in nonwhite_positions]),
+            min([y for x, y in nonwhite_positions]),
+            max([x for x, y in nonwhite_positions]),
+            max([y for x, y in nonwhite_positions]),
+        )
+        return ex_21_img.crop(rect)
```

Comment:
I'm a little worried about the resolution of the saved crops, as it seems much lower than a screenshot of the rendered HTML. It might not be an issue after the model preprocesses, resizes, and creates an embedding of the image. I'll try labeling and passing one of the images through the model and see what happens, but just flagging that the resolution seems kind of low.

Also, I think it's likely that we'll need some whitespace padding around the table for the model to work best, so maybe this cropping is something to revisit later anyways.